csrc/host_ir/ir.h

```diff
@@ -150,6 +155,13 @@ class LaunchKernel : public Expr {
   Val* cacheId() const {
     return attributeVal(3);
   }
+
+  CompiledKernel* compiledKernel() const {
+    return compiled_kernel_;
+  }
+
+ private:
+  CompiledKernel* compiled_kernel_ = nullptr;
```

Comment:
In general, `Expr` subclasses should not have their own members like this but hold additional data in `Expr::attributes_`. Some of the Expr utilities such as `shallowCopy` wouldn't work otherwise. 

Holding `CompiledKernel` in the attribute list requires `CompiledKernel` to be a `Val`. That at least conceptually makes sense, although implementation wise may make things more complicated.

Reply:
After discussion with @wujingyue the correct solution for this is going to require a more targeted refactor around compilation responsibility and the ownership of `CompiledKernel`. Since this PR is getting rather large I'm opening a separate issue #5652  for this and will follow up on this with a dedicated change.